miri test is painful to use with CARGO_TARGET_DIR set

I have `CARGO_TARGET_DIR` set on my machine to shared the target directories of the many rust projects on my computer. This generally works very well, and reduces build times (often by a lot!) by re-using build artifacts across crates I'm working on. Unfortunately, `miri test` _also_ shares that build directory, and when it re-compiles crates with a modified `std`, it causes non-miri builds to fail with errors like:

```
error[E0460]: found possibly newer version of crate `core` which `lazy_static` depends on
```

This is solved easily enough by remembering to always run miri test with:
```console
$ env -u CARGO_TARGET_DIR cargo miri test
```
But, well, I often forget, and then have to wipe my whole shared target directory, which leads to very long compile times the next time I work on a large project.

I wonder if it might be possible to have `miri` use it's own "build profile". `rls` already does this (its artifacts ends up with `target/rls`), and if miri also did that (`target/miri`), all of these issues would simply go away! I don't know what mechanisms would be involved, but it would save me (and probably others) a lot of headache by eliminating the error above :)

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -74,6 +74,7 @@ jobs:
           else
             RUSTC_HASH=$(< rust-version)
           fi
+          # We need a nightly cargo for parts of the cargo miri test suite.
           rustup-toolchain-install-master \
             -f \
             -n master "$RUSTC_HASH" \

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -83,6 +83,12 @@ jobs:
             --host ${{ matrix.host_target }}
           rustup default master
 
+      # We need a nightly Cargo to run tests that depend on unstable Cargo features.
+      - name: Install latest nightly
+        uses: actions-rs/toolchain@v1
+        with:
+          toolchain: nightly
+
       - name: Show Rust version
         run: |
           rustup show

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -77,17 +77,12 @@ jobs:
           rustup-toolchain-install-master \
             -f \
             -n master "$RUSTC_HASH" \
+            -c cargo \
             -c rust-src \
             -c rustc-dev \
             -c llvm-tools \
             --host ${{ matrix.host_target }}
           rustup default master
-
-      # We need a nightly Cargo to run tests that depend on unstable Cargo features.
-      - name: Install latest nightly
-        uses: actions-rs/toolchain@v1
-        with:
-          toolchain: nightly
 
       - name: Show Rust version
         run: |

--- a/cargo-miri/bin.rs
+++ b/cargo-miri/bin.rs
@@ -6,7 +6,7 @@ use std::io::{self, BufRead, BufReader, BufWriter, Read, Write};
 use std::iter::TakeWhile;
 use std::ops::Not;
 use std::path::{Path, PathBuf};
-use std::process::{Command, Stdio};
+use std::process::{self, Command};
 
 use serde::{Deserialize, Serialize};
 
@@ -233,7 +233,7 @@ fn exec(mut cmd: Command) {
 /// If it fails, fail this process with the same exit code.
 /// Otherwise, continue.
 fn exec_with_pipe(mut cmd: Command, input: &[u8]) {
-    cmd.stdin(std::process::Stdio::piped());
+    cmd.stdin(process::Stdio::piped());
     let mut child = cmd.spawn().expect("failed to spawn process");
     {
         let stdin = child.stdin.as_mut().expect("failed to open stdin");
@@ -493,8 +493,8 @@ fn detect_target_dir() -> PathBuf {
         }
     }
     let mut child = cmd
-        .stdin(Stdio::null())
-        .stdout(Stdio::piped())
+        .stdin(process::Stdio::null())
+        .stdout(process::Stdio::piped())
         .spawn()
         .expect("failed ro run `cargo metadata`");
     // Check this `Result` after `status.success()` is checked, so we don't print the error

--- a/cargo-miri/bin.rs
+++ b/cargo-miri/bin.rs
@@ -500,7 +500,7 @@ fn detect_target_dir() -> PathBuf {
     // Check this `Result` after `status.success()` is checked, so we don't print the error
     // to stderr if `cargo metadata` is also printing to stderr.
     let metadata: Result<Metadata, _> = serde_json::from_reader(child.stdout.take().unwrap());
-    let status = child.wait().expect("failed to wait `cargo metadata` to exit");
+    let status = child.wait().expect("failed to wait for `cargo metadata` to exit");
     if !status.success() {
         std::process::exit(status.code().unwrap_or(-1));
     }

--- a/cargo-miri/bin.rs
+++ b/cargo-miri/bin.rs
@@ -572,7 +572,12 @@ fn phase_cargo_miri(mut args: env::Args) {
     // Forward all arguments before `--` other than `--target-dir` and its value to Cargo.
     for arg in ArgSplitFlagValue::new(&mut args, "--target-dir") {
         match arg {
-            Ok(value) => target_dir = Some(value.into()),
+            Ok(value) => {
+                if target_dir.is_some() {
+                    show_error(format!("`--target-dir` is provided more than once"));
+                }
+                target_dir = Some(value.into());
+            }
             Err(arg) => drop(cmd.arg(arg)),
         }
     }

--- a/cargo-miri/bin.rs
+++ b/cargo-miri/bin.rs
@@ -578,7 +578,9 @@ fn phase_cargo_miri(mut args: env::Args) {
                 }
                 target_dir = Some(value.into());
             }
-            Err(arg) => drop(cmd.arg(arg)),
+            Err(arg) => {
+                cmd.arg(arg);
+            }
         }
     }
 

--- a/cargo-miri/bin.rs
+++ b/cargo-miri/bin.rs
@@ -114,12 +114,12 @@ fn has_arg_flag(name: &str) -> bool {
 
 /// Yields all values of command line flag `name` as `Ok(arg)`, and all other arguments except
 /// the flag as `Err(arg)`.
-struct ArgFlagValueWithOtherArgsIter<'a, I> {
+struct ArgSplitFlagValue<'a, I> {
     args: TakeWhile<I, fn(&String) -> bool>,
     name: &'a str,
 }
 
-impl<'a, I: Iterator<Item = String>> ArgFlagValueWithOtherArgsIter<'a, I> {
+impl<'a, I: Iterator<Item = String>> ArgSplitFlagValue<'a, I> {
     fn new(args: I, name: &'a str) -> Self {
         Self {
             // Stop searching at `--`.
@@ -129,7 +129,7 @@ impl<'a, I: Iterator<Item = String>> ArgFlagValueWithOtherArgsIter<'a, I> {
     }
 }
 
-impl<I: Iterator<Item = String>> Iterator for ArgFlagValueWithOtherArgsIter<'_, I> {
+impl<I: Iterator<Item = String>> Iterator for ArgSplitFlagValue<'_, I> {
     type Item = Result<String, String>;
 
     fn next(&mut self) -> Option<Self::Item> {
@@ -151,11 +151,11 @@ impl<I: Iterator<Item = String>> Iterator for ArgFlagValueWithOtherArgsIter<'_, 
 }
 
 /// Yields all values of command line flag `name`.
-struct ArgFlagValueIter<'a>(ArgFlagValueWithOtherArgsIter<'a, env::Args>);
+struct ArgFlagValueIter<'a>(ArgSplitFlagValue<'a, env::Args>);
 
 impl<'a> ArgFlagValueIter<'a> {
     fn new(name: &'a str) -> Self {
-        Self(ArgFlagValueWithOtherArgsIter::new(env::args(), name))
+        Self(ArgSplitFlagValue::new(env::args(), name))
     }
 }
 
@@ -484,7 +484,7 @@ fn detect_target_dir() -> PathBuf {
     // The `build.target-dir` config can by passed by `--config` flags, so forward them to
     // `cargo metadata`.
     let config_flag = "--config";
-    for arg in ArgFlagValueWithOtherArgsIter::new(
+    for arg in ArgSplitFlagValue::new(
         env::args().skip(3), // skip the program name, "miri" and "run" / "test"
         config_flag,
     ) {
@@ -570,7 +570,7 @@ fn phase_cargo_miri(mut args: env::Args) {
     let mut target_dir = None;
 
     // Forward all arguments before `--` other than `--target-dir` and its value to Cargo.
-    for arg in ArgFlagValueWithOtherArgsIter::new(&mut args, "--target-dir") {
+    for arg in ArgSplitFlagValue::new(&mut args, "--target-dir") {
         match arg {
             Ok(value) => target_dir = Some(value.into()),
             Err(arg) => drop(cmd.arg(arg)),

--- a/cargo-miri/bin.rs
+++ b/cargo-miri/bin.rs
@@ -113,7 +113,7 @@ fn has_arg_flag(name: &str) -> bool {
 }
 
 /// Yields all values of command line flag `name` as `Ok(arg)`, and all other arguments except
-/// the flag as `Err(arg)`.
+/// the flag as `Err(arg)`. (The flag `name` itself is not yielded at all, only its values are.)
 struct ArgSplitFlagValue<'a, I> {
     args: TakeWhile<I, fn(&String) -> bool>,
     name: &'a str,

--- a/cargo-miri/bin.rs
+++ b/cargo-miri/bin.rs
@@ -6,7 +6,7 @@ use std::io::{self, BufRead, BufReader, BufWriter, Read, Write};
 use std::iter::TakeWhile;
 use std::ops::Not;
 use std::path::{Path, PathBuf};
-use std::process::Command;
+use std::process::{Command, Stdio};
 
 use serde::{Deserialize, Serialize};
 
@@ -112,19 +112,50 @@ fn has_arg_flag(name: &str) -> bool {
     args.any(|val| val == name)
 }
 
-/// Yields all values of command line flag `name`.
-struct ArgFlagValueIter<'a> {
-    args: TakeWhile<env::Args, fn(&String) -> bool>,
+/// Yields all values of command line flag `name` as `Ok(arg)`, and all other arguments except
+/// the flag as `Err(arg)`.
+struct ArgFlagValueWithOtherArgsIter<'a, I> {
+    args: TakeWhile<I, fn(&String) -> bool>,
     name: &'a str,
 }
 
-impl<'a> ArgFlagValueIter<'a> {
-    fn new(name: &'a str) -> Self {
+impl<'a, I: Iterator<Item = String>> ArgFlagValueWithOtherArgsIter<'a, I> {
+    fn new(args: I, name: &'a str) -> Self {
         Self {
             // Stop searching at `--`.
-            args: env::args().take_while(|val| val != "--"),
+            args: args.take_while(|val| val != "--"),
             name,
         }
+    }
+}
+
+impl<I: Iterator<Item = String>> Iterator for ArgFlagValueWithOtherArgsIter<'_, I> {
+    type Item = Result<String, String>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let arg = self.args.next()?;
+        if arg.starts_with(self.name) {
+            // Strip leading `name`.
+            let suffix = &arg[self.name.len()..];
+            if suffix.is_empty() {
+                // This argument is exactly `name`; the next one is the value.
+                return self.args.next().map(Ok);
+            } else if suffix.starts_with('=') {
+                // This argument is `name=value`; get the value.
+                // Strip leading `=`.
+                return Some(Ok(suffix[1..].to_owned()));
+            }
+        }
+        Some(Err(arg))
+    }
+}
+
+/// Yields all values of command line flag `name`.
+struct ArgFlagValueIter<'a>(ArgFlagValueWithOtherArgsIter<'a, env::Args>);
+
+impl<'a> ArgFlagValueIter<'a> {
+    fn new(name: &'a str) -> Self {
+        Self(ArgFlagValueWithOtherArgsIter::new(env::args(), name))
     }
 }
 
@@ -133,19 +164,8 @@ impl Iterator for ArgFlagValueIter<'_> {
 
     fn next(&mut self) -> Option<Self::Item> {
         loop {
-            let arg = self.args.next()?;
-            if !arg.starts_with(self.name) {
-                continue;
-            }
-            // Strip leading `name`.
-            let suffix = &arg[self.name.len()..];
-            if suffix.is_empty() {
-                // This argument is exactly `name`; the next one is the value.
-                return self.args.next();
-            } else if suffix.starts_with('=') {
-                // This argument is `name=value`; get the value.
-                // Strip leading `=`.
-                return Some(suffix[1..].to_owned());
+            if let Ok(value) = self.0.next()? {
+                return Some(value);
             }
         }
     }
@@ -510,8 +530,59 @@ fn phase_cargo_miri(mut args: env::Args) {
         &host
     };
 
-    // Forward all further arguments to cargo.
-    cmd.args(args);
+    let mut target_dir = None;
+
+    // Forward all arguments before `--` other than `--target-dir` and its value to Cargo.
+    for arg in ArgFlagValueWithOtherArgsIter::new(&mut args, "--target-dir") {
+        match arg {
+            Ok(value) => target_dir = Some(value.into()),
+            Err(arg) => drop(cmd.arg(arg)),
+        }
+    }
+
+    // Detect the target directory if it's not specified via `--target-dir`.
+    let target_dir = target_dir.get_or_insert_with(|| {
+        #[derive(Deserialize)]
+        struct Metadata {
+            target_directory: PathBuf,
+        }
+        let mut cmd = cargo();
+        // `-Zunstable-options` is required by `--config`.
+        cmd.args(["metadata", "--no-deps", "--format-version=1", "-Zunstable-options"]);
+        // The `build.target-dir` config can by passed by `--config` flags, so forward them to
+        // `cargo metadata`.
+        let config_flag = "--config";
+        for arg in ArgFlagValueWithOtherArgsIter::new(
+            env::args().skip(3), // skip the program name, "miri" and "run" / "test"
+            config_flag,
+        ) {
+            if let Ok(config) = arg {
+                cmd.arg(config_flag).arg(config);
+            }
+        }
+        let mut child = cmd
+            .stdin(Stdio::null())
+            .stdout(Stdio::piped())
+            .spawn()
+            .expect("failed ro run `cargo metadata`");
+        // Check this `Result` after `status.success()` is checked, so we don't print the error
+        // to stderr if `cargo metadata` is also printing to stderr.
+        let metadata: Result<Metadata, _> = serde_json::from_reader(child.stdout.take().unwrap());
+        let status = child.wait().expect("failed to wait `cargo metadata` to exit");
+        if !status.success() {
+            std::process::exit(status.code().unwrap_or(-1));
+        }
+        metadata
+            .unwrap_or_else(|e| show_error(format!("invalid `cargo metadata` output: {}", e)))
+            .target_directory
+    });
+
+    // Set `--target-dir` to `miri` inside the original target directory.
+    target_dir.push("miri");
+    cmd.arg("--target-dir").arg(target_dir);
+
+    // Forward all further arguments after `--` to cargo.
+    cmd.arg("--").args(args);
 
     // Set `RUSTC_WRAPPER` to ourselves.  Cargo will prepend that binary to its usual invocation,
     // i.e., the first argument is `rustc` -- which is what we use in `main` to distinguish

--- a/cargo-miri/bin.rs
+++ b/cargo-miri/bin.rs
@@ -472,6 +472,43 @@ path = "lib.rs"
     }
 }
 
+/// Detect the target directory by calling `cargo metadata`.
+fn detect_target_dir() -> PathBuf {
+    #[derive(Deserialize)]
+    struct Metadata {
+        target_directory: PathBuf,
+    }
+    let mut cmd = cargo();
+    // `-Zunstable-options` is required by `--config`.
+    cmd.args(["metadata", "--no-deps", "--format-version=1", "-Zunstable-options"]);
+    // The `build.target-dir` config can by passed by `--config` flags, so forward them to
+    // `cargo metadata`.
+    let config_flag = "--config";
+    for arg in ArgFlagValueWithOtherArgsIter::new(
+        env::args().skip(3), // skip the program name, "miri" and "run" / "test"
+        config_flag,
+    ) {
+        if let Ok(config) = arg {
+            cmd.arg(config_flag).arg(config);
+        }
+    }
+    let mut child = cmd
+        .stdin(Stdio::null())
+        .stdout(Stdio::piped())
+        .spawn()
+        .expect("failed ro run `cargo metadata`");
+    // Check this `Result` after `status.success()` is checked, so we don't print the error
+    // to stderr if `cargo metadata` is also printing to stderr.
+    let metadata: Result<Metadata, _> = serde_json::from_reader(child.stdout.take().unwrap());
+    let status = child.wait().expect("failed to wait `cargo metadata` to exit");
+    if !status.success() {
+        std::process::exit(status.code().unwrap_or(-1));
+    }
+    metadata
+        .unwrap_or_else(|e| show_error(format!("invalid `cargo metadata` output: {}", e)))
+        .target_directory
+}
+
 fn phase_cargo_miri(mut args: env::Args) {
     // Check for version and help flags even when invoked as `cargo-miri`.
     if has_arg_flag("--help") || has_arg_flag("-h") {
@@ -541,41 +578,7 @@ fn phase_cargo_miri(mut args: env::Args) {
     }
 
     // Detect the target directory if it's not specified via `--target-dir`.
-    let target_dir = target_dir.get_or_insert_with(|| {
-        #[derive(Deserialize)]
-        struct Metadata {
-            target_directory: PathBuf,
-        }
-        let mut cmd = cargo();
-        // `-Zunstable-options` is required by `--config`.
-        cmd.args(["metadata", "--no-deps", "--format-version=1", "-Zunstable-options"]);
-        // The `build.target-dir` config can by passed by `--config` flags, so forward them to
-        // `cargo metadata`.
-        let config_flag = "--config";
-        for arg in ArgFlagValueWithOtherArgsIter::new(
-            env::args().skip(3), // skip the program name, "miri" and "run" / "test"
-            config_flag,
-        ) {
-            if let Ok(config) = arg {
-                cmd.arg(config_flag).arg(config);
-            }
-        }
-        let mut child = cmd
-            .stdin(Stdio::null())
-            .stdout(Stdio::piped())
-            .spawn()
-            .expect("failed ro run `cargo metadata`");
-        // Check this `Result` after `status.success()` is checked, so we don't print the error
-        // to stderr if `cargo metadata` is also printing to stderr.
-        let metadata: Result<Metadata, _> = serde_json::from_reader(child.stdout.take().unwrap());
-        let status = child.wait().expect("failed to wait `cargo metadata` to exit");
-        if !status.success() {
-            std::process::exit(status.code().unwrap_or(-1));
-        }
-        metadata
-            .unwrap_or_else(|e| show_error(format!("invalid `cargo metadata` output: {}", e)))
-            .target_directory
-    });
+    let target_dir = target_dir.get_or_insert_with(detect_target_dir);
 
     // Set `--target-dir` to `miri` inside the original target directory.
     target_dir.push("miri");

--- a/cargo-miri/bin.rs
+++ b/cargo-miri/bin.rs
@@ -481,7 +481,7 @@ fn detect_target_dir() -> PathBuf {
     let mut cmd = cargo();
     // `-Zunstable-options` is required by `--config`.
     cmd.args(["metadata", "--no-deps", "--format-version=1", "-Zunstable-options"]);
-    // The `build.target-dir` config can by passed by `--config` flags, so forward them to
+    // The `build.target-dir` config can be passed by `--config` flags, so forward them to
     // `cargo metadata`.
     let config_flag = "--config";
     for arg in ArgSplitFlagValue::new(

--- a/rustup-toolchain
+++ b/rustup-toolchain
@@ -39,7 +39,7 @@ fi
 
 # Install and setup new toolchain.
 rustup toolchain uninstall miri
-rustup-toolchain-install-master -n miri -c rust-src -c rustc-dev -c llvm-tools -- "$NEW_COMMIT"
+rustup-toolchain-install-master -n miri -c cargo -c rust-src -c rustc-dev -c llvm-tools -- "$NEW_COMMIT"
 rustup override set miri
 
 # Cleanup.

--- a/test-cargo-miri/.gitignore
+++ b/test-cargo-miri/.gitignore
@@ -1,1 +1,4 @@
 *.real
+custom-run
+custom-test
+config-cli

--- a/test-cargo-miri/run-test.py
+++ b/test-cargo-miri/run-test.py
@@ -173,9 +173,11 @@ if not 'MIRI_SYSROOT' in os.environ:
     subprocess.run(cargo_miri("setup"), check=True)
 test_cargo_miri_run()
 test_cargo_miri_test()
+# Ensure we did not create anything outside the expected target dir.
 for target_dir in ["target", "custom-run", "custom-test", "config-cli"]:
     if os.listdir(target_dir) != ["miri"]:
         fail(f"`{target_dir}` contains unexpected files")
+    # Ensure something exists inside that target dir.
     os.access(os.path.join(target_dir, "miri", "debug", "deps"), os.F_OK)
 
 print("\nTEST SUCCESSFUL!")

--- a/test-cargo-miri/run-test.py
+++ b/test-cargo-miri/run-test.py
@@ -101,6 +101,11 @@ def test_cargo_miri_run():
         "run.subcrate.stdout.ref", "run.subcrate.stderr.ref",
         env={'MIRIFLAGS': "-Zmiri-disable-isolation"},
     )
+    test("`cargo miri run` (custom target dir)",
+        # Attempt to confuse the argument parser.
+        cargo_miri("run") + ["--target-dir=custom-run", "--", "--target-dir=target/custom-run"],
+        "run.args.stdout.ref", "run.custom-target-dir.stderr.ref",
+    )
 
 def test_cargo_miri_test():
     # rustdoc is not run on foreign targets
@@ -144,8 +149,18 @@ def test_cargo_miri_test():
         cargo_miri("test") + ["-p", "subcrate", "--doc"],
         "test.stdout-empty.ref", "test.stderr-proc-macro-doctest.ref",
     )
+    test("`cargo miri test` (custom target dir)",
+        cargo_miri("test") + ["--target-dir=custom-test"],
+        default_ref, "test.stderr-empty.ref",
+    )
+    del os.environ["CARGO_TARGET_DIR"] # this overrides `build.target-dir` passed by `--config`, so unset it
+    test("`cargo miri test` (config-cli)",
+        cargo_miri("test") + ["--config=build.target-dir=\"config-cli\"", "-Zunstable-options"],
+        default_ref, "test.stderr-empty.ref",
+    )
 
 os.chdir(os.path.dirname(os.path.realpath(__file__)))
+os.environ["CARGO_TARGET_DIR"] = "target" # this affects the location of the target directory that we need to check
 os.environ["RUST_TEST_NOCAPTURE"] = "0" # this affects test output, so make sure it is not set
 os.environ["RUST_TEST_THREADS"] = "1" # avoid non-deterministic output due to concurrent test runs
 
@@ -158,6 +173,10 @@ if not 'MIRI_SYSROOT' in os.environ:
     subprocess.run(cargo_miri("setup"), check=True)
 test_cargo_miri_run()
 test_cargo_miri_test()
+for target_dir in ["target", "custom-run", "custom-test", "config-cli"]:
+    if os.listdir(target_dir) != ["miri"]:
+        fail(f"`{target_dir}` contains unexpected files")
+    os.access(os.path.join(target_dir, "miri", "debug", "deps"), os.F_OK)
 
 print("\nTEST SUCCESSFUL!")
 sys.exit(0)

--- a/test-cargo-miri/run.custom-target-dir.stderr.ref
+++ b/test-cargo-miri/run.custom-target-dir.stderr.ref
@@ -1,0 +1,2 @@
+main
+--target-dir=target/custom-run

